packages/bff/src/services/toolsService.ts

```diff
@@ -0,0 +1,393 @@
+        });
+      }
+
+      const { errors: platformStateErrors } =
+        validateClientKindAndPlatformState(key, jwt);
+
+      if (platformStateErrors) {
+        return handleValidationResults({
+          platformStateErrors,
+        });
+      }
+
+      return handleValidationResults({});
+    },
+  };
+}
+
+export type ToolsService = ReturnType<typeof toolsServiceBuilder>;
+
+function handleValidationResults(
+  errs: {
+    clientAssertionErrors?: Array<ApiError<string>>;
+    keyRetrieveErrors?: Array<ApiError<string>>;
+    clientSignatureErrors?: Array<ApiError<string>>;
+    platformStateErrors?: Array<ApiError<string>>;
+  },
+  clientKind?: authorizationApi.ClientKind,
+  eservice?: bffApi.TokenGenerationValidationEService
+): bffApi.TokenGenerationValidationResult {
+  const clientAssertionErrors = errs.clientAssertionErrors ?? [];
+  const keyRetrieveErrors = errs.keyRetrieveErrors ?? [];
+  const clientSignatureErrors = errs.clientSignatureErrors ?? [];
+  const platformStateErrors = errs.platformStateErrors ?? [];
+
+  return {
+    clientKind,
+    eservice,
+    steps: {
+      clientAssertionValidation: {
+        result: getStepResult([], clientAssertionErrors),
+        failures: apiErrorsToValidationFailures(clientAssertionErrors),
+      },
+      publicKeyRetrieve: {
+        result: getStepResult(clientAssertionErrors, keyRetrieveErrors),
+        failures: apiErrorsToValidationFailures(keyRetrieveErrors),
+      },
+      clientAssertionSignatureVerification: {
+        result: getStepResult(
+          [...clientAssertionErrors, ...keyRetrieveErrors],
+          clientSignatureErrors
+        ),
+        failures: apiErrorsToValidationFailures(clientSignatureErrors),
+      },
+      platformStatesVerification: {
+        result: getStepResult(
+          [
+            ...clientAssertionErrors,
+            ...keyRetrieveErrors,
+            ...clientSignatureErrors,
+          ],
+          platformStateErrors
+        ),
+        failures: apiErrorsToValidationFailures(platformStateErrors),
+      },
+    },
+  };
+}
+
+function getStepResult(
+  prevStepErrors: Array<ApiError<string>>,
+  currentStepErrors: Array<ApiError<string>>
+): bffApi.TokenGenerationValidationStepResult {
+  if (currentStepErrors.length > 0) {
+    return bffApi.TokenGenerationValidationStepResult.Enum.FAILED;
+  } else if (prevStepErrors.length > 0) {
+    return bffApi.TokenGenerationValidationStepResult.Enum.SKIPPED;
+  } else {
+    return bffApi.TokenGenerationValidationStepResult.Enum.PASSED;
+  }
+}
+
+async function retrieveKey(
+  {
+    authorizationClient,
+    purposeProcessClient,
+    agreementProcessClient,
+    catalogProcessClient,
+  }: PagoPAInteropBeClients,
+  jwt: ClientAssertion,
+  ctx: WithLogger<BffAppContext>
+): Promise<
+  SuccessfulValidation<ApiKey | ConsumerKey> | FailedValidation<ErrorCodes>
+> {
+  const client = await authorizationClient.token
+    .getKeyWithClientByKeyId({
```

Comment:
In Scala this API call is not performed to the authorizationProcess, but to the authorizationManagement - was this change discussed? Maybe the result is the same @galales?